### PR TITLE
EPMRPP-50911 Parallel run support

### PR DESCRIPTION
**Pull request by AmsterGet**
* Added separate example for parallel execution with RealTimeReporter
* IPC used instead of Node.js events emitter

Scheme for parallel run (see the screenshot below).

image

### CODE REVIEW

File: examples/RTR-exampleChromeDriver/tests/home.js [hunk 1]
```diff
@@ -6,39 +6,36 @@ describe(suiteName, function() {
 
   this.retries(3);
 
-  before((browser, done) => {
+  before(() => {
     const item = {
       name: suiteName,
-      attributes: [{ key: 'suite', value: 'home' }],
-      description: 'Common suite description',
+      attributes: [{ key: 'suite', value: 'home.js' }],
```
Comment (KatsiarynaTatarynovich) [inline examples/RTR-exampleChromeDriver/tests/home.js:12]: Could you please tell me why you added the file format? Does it carry some meaning?

File: src/realTimeReporter/ipc/client.ts [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+ *  limitations under the License.
+ *
+ */
+
+import ipc from 'node-ipc';
+import { EVENTS } from '../../constants';
+
+export const connectIPCClient = () => {
+  ipc.config.id = `reportPortalReporter_${process.pid}`;
+  ipc.config.retry = 1500;
```
Comment (KatsiarynaTatarynovich) [inline src/realTimeReporter/ipc/client.ts:23]: Why exactly this value? `1500`

File: examples/RTR-exampleChromeDriver/tests/home.js (hunk 1 above)
Comment (AmsterGet) [reply examples/RTR-exampleChromeDriver/tests/home.js:12]: It's just an example. So it's not important, I think :)

File: src/realTimeReporter/ipc/client.ts (hunk 2 above)
Comment (AmsterGet) [reply src/realTimeReporter/ipc/client.ts:23]: This value is obtained experimentally. This time is enough to exclude the possibility of a client connecting to the server before it starts.

**Comment by ElenaRomanchuk** [comment]
Also why do you add description instead of set and overwrite it? As I remember, other JS-agents just set description, not append.

File: examples/RTR-exampleChromeDriverParallelRun/tests/home.js [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+  before(() => {
+    const item = {
+      name: suiteName,
+      attributes: [{ key: 'suite', value: 'home' }],
+      description: 'Common suite description',
+    };
+    PublicReportingAPI.startSuite(item);
+
+    PublicReportingAPI.startBeforeSuite(suiteName);
```
[the remark this answers is not in the document]
Comment (AmsterGet) [reply examples/RTR-exampleChromeDriverParallelRun/tests/home.js:35]: In nightwatch every test file == test suite.
So here if we want to report beforeSuite hook to a step level we should do it after the parent item start request.
In other case this item will be on the suite level. Is it wrong?
Shouldn't before/after hooks be inside the suite, but before/after all test executuions?

[the remark this answers is not in the document]
Comment (AmsterGet) [reply examples/RTR-exampleChromeDriverParallelRun/tests/home.js:35]: We decided to leave the hooks at the same level as the test item.

File: src/realTimeReporter/reportingApi/attachData.ts [hunk 4]
```diff
@@ -15,85 +15,126 @@
...
-        PublicReportingAPI.addLaunchLog({ level: LOG_LEVELS.FATAL, message, file });
-    },
+  addAttributes(attributes, itemName) {
+    publishEvent(EVENTS.ADD_ATTRIBUTES, { attributes, itemName });
+  },
+  addDescription(text, itemName) {
+    publishEvent(EVENTS.ADD_DESCRIPTION, { text, itemName });
+  },
+
+  log(level, message, file = null, itemName) {
```
[the remark this answers is not in the document]
Comment (AmsterGet) [reply src/realTimeReporter/reportingApi/attachData.ts:54]: Purpose is unknown) Removed.

File: src/realTimeReporter/ipc/server.ts [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+
+import ipc from 'node-ipc';
+
+export const startIPCServer = (subscribeServerEvents: (server: any) => void) => {
+    if (ipc.server) {
+        subscribeServerEvents(ipc.server);
```
[the remark this answers is not in the document]
Comment (AmsterGet) [reply src/realTimeReporter/ipc/server.ts:22]: Agreed. Fixed.

**Pull request merged**